### ISSUE TITLE
searborn adaptation need for Matptotlib-2.2.0rc1

### ISSUE THREAD

**Issue by stonebig** [matplotlib/matplotlib#10585]
### Bug report

**Bug summary**

My historic Seaborn example doesn't seem to work anymore with Matplotlib-2.2.0rc1. is it a volontary change ?

```python
# Seaborn
# for more examples, see http://stanford.edu/~mwaskom/software/seaborn/examples/index.html
import seaborn as sns
sns.set()
df = sns.load_dataset("iris")
sns.pairplot(df, hue="species", size=1.5)
```

**Actual outcome**

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-78e2f3db95aa> in <module>()
      4 sns.set()
      5 df = sns.load_dataset("iris")
----> 6 sns.pairplot(df, hue="species", size=1.5)

C:\WinPython\bd35\buildQt5\winpython-32bit-3.5.x.2\python-3.5.4\lib\site-packages\seaborn\axisgrid.py in pairplot(data, hue, hue_order, palette, vars, x_vars, y_vars, kind, diag_kind, markers, size, aspect, dropna, plot_kws, diag_kws, grid_kws)
   2058     if grid.square_grid:
   2059         if diag_kind == "hist":
-> 2060             grid.map_diag(plt.hist, **diag_kws)
   2061         elif diag_kind == "kde":
   2062             diag_kws["legend"] = False

C:\WinPython\bd35\buildQt5\winpython-32bit-3.5.x.2\python-3.5.4\lib\site-packages\seaborn\axisgrid.py in map_diag(self, func, **kwargs)
   1333                     diag_ax = ax._make_twin_axes(sharex=ax,
   1334                                                  sharey=diag_axes[0],
-> 1335                                                  frameon=False)
   1336                 else:
   1337                     diag_ax = ax._make_twin_axes(sharex=ax, frameon=False)

C:\WinPython\bd35\buildQt5\winpython-32bit-3.5.x.2\python-3.5.4\lib\site-packages\matplotlib\axes\_subplots.py in _make_twin_axes(self, *kl, **kwargs)
    174         from matplotlib.projections import process_projection_requirements
    175         if 'sharex' in kwargs and 'sharey' in kwargs:
--> 176             raise ValueError("Twinned Axes may share only one axis.")
    177         kl = (self.get_subplotspec(),) + kl
    178         projection_class, kwargs, key = process_projection_requirements(

ValueError: Twinned Axes may share only one axis.
```

**Matplotlib version**

  * Operating system: windows10
  * Matplotlib version:  2.2.0rc1 of feb 18th cgohlke
  * Matplotlib backend (`print(matplotlib.get_backend())`):module://ipykernel.pylab.backend_inline
  * Python version: 3.5.4
  * Jupyter version (if applicable): Jupyterlab-0.31.8
  * Other libraries:

**Comment by stonebig** [comment]
so it comes from this change:
https://github.com/matplotlib/matplotlib/pull/10033

how should we rewrite an old code like :
````
ax._make_twin_axes(sharex=ax,
 sharey=diag_axes[0],
 frameon=False)
````

**Comment by tacaswell** [comment]
attn @efiring 

I think was was intentional to account for interactions between twin, shared axes, and fixed aspect ratio.

**Comment by stonebig** [comment]
ok, I patched seaborn for my need. I understand it was an internal API.

**Comment by jklymak** [comment]
Can you distill to a non-seaborn example?  I’m still not sure what seaborn is trying to do that breaks, but I haven’t read the source.

**Comment by stonebig** [comment]
Similar examples would come from holoviews (replace 'bokeh' per 'matplotlib'):
- http://holoviews.org/gallery/demos/bokeh/iris_splom_example.html#bokeh-gallery-iris-splom-example

**Comment by stonebig** [comment]
more  similar woub be with hv.Histogram
````
from bokeh.sampledata.iris import flowers
from holoviews.operation import gridmatrix
ds = hv.Dataset(flowers)
grouped_by_species = ds.groupby('species', container_type=hv.NdOverlay)
grid = gridmatrix(grouped_by_species, diagonal_type=hv.Histogram)
````

**Comment by jklymak** [comment]
I meant can you trigger the error using matplotlib only?

**Comment by stonebig** [comment]
hum, it's not an internal error of Matplotlib-2.2.0rc1, if that is your question.

**Comment by jklymak** [comment]
So I'm confused; is this a matplotlib problem (i.e. something that worked pre #10033)?  If so, then it must be triggerable with just matplotlib calls.

**Comment by stonebig** [comment]
seabor was using an "internal API", functions starting per the undescore characters:
-  there has been a change there,
- and seaborn is no longer actively maintained.

**Comment by efiring** [comment]
The change in `_make_twin_axes` was deliberate.  Under what circumstances does it make sense to share both axes with twin axes?  The point of twin axes is to share one axis and have the other one differ.

**Comment by stonebig** [comment]
ok. I agree it was probably a functional bug of seaborn.

**Comment by mwaskom** [comment]
@efiring 

> The change in _make_twin_axes was deliberate. Under what circumstances does it make sense to share both axes with twin axes? The point of twin axes is to share one axis and have the other one differ.

seaborn shouldn't have been using a private method here, but if the issue is just redundancy, is there any reason matplotlib needs to raise on getting `sharex` and `sharey` when one of the shared axes is self?

**Comment by mwaskom** [comment]
Here's a self-contained example of what works on 2.1.1 and I gather fails on 2.2:

```python
f, (ax0, ax1) = plt.subplots(1, 2, sharey=True)
ax2 = ax1._make_twin_axes(sharex=ax1, sharey=ax0)
```

**Comment by mwaskom** [comment]
BTW the reason seaborn was using the private method is that I don't believe it is possible through the public API to create a shared twinned axes (at least, `ax.twinx` has no parameters that would allow this, maybe there is another method).

**Comment by efiring** [comment]
I don't yet understand what you are trying to do in that example.  It looks like ax1 and ax2, the twins with their shared x-axis, are being forced to have the same y-axis as well, via the sharing with ax0.  In that case, what is the need for having both ax1 and ax2?  Their x axes are identical, and their y axes are identical, so why do you need two Axes?

Adding the check to `_make_twin_axes` was part of an attempt to prevent over-specification or inconsistent specifications of sharing, twinning (which locks the two position boxes together) and aspect-ratio control.  I'm sure it is still not optimal.

**Comment by mwaskom** [comment]
> I don't yet understand what you are trying to do in that example. It looks like ax1 and ax2, the twins with their shared x-axis, are being forced to have the same y-axis as well, via the sharing with ax0. So why do you need two Axes?

The actual use case is a bit more complicated. The goal is to make a home for the for the histograms on the diagonal axes of a seaborn [`PairGrid`](http://seaborn.pydata.org/examples/scatterplot_matrix.html) (aka a scatterplot matrix). The point is that the grid is set up so that x axes are shared across rows and y axes are shared across columns. The histograms are drawn on twins of the diagonal axes so that they share x axes with the rest of the plots in each column, but their y axes are shared *across the diagonal* to show the counts on the same scale. Does that make sense?

> Adding the check to _make_twin_axes was part of an attempt to prevent over-specification or inconsistent specifications of sharing, twinning (which locks the two position boxes together) and aspect-ratio control. I'm sure it is still not optimal.

As I see it, the error here is a false positive because ax2 really does share one axis with its twin. It's raised because `sharex` is explicitly specified where otherwise you'd get the same result implicitly. I think it should be easy to modify the checked condition not to raise if both `sharex` and `sharey` are passed but one of them is `self`.

Granted, this is a private API, so you can feel free to change it at will, but if an easy fix could avoid completely breaking a popular seaborn function, that would be rad.

http://seaborn.pydata.org/examples/scatterplot_matrix.html

**Comment by jklymak** [comment]
If this can't be achieved w/ the public API, then we should develop a public API to do it, and have tests etc.  

@mwaskom can you make a 2x2 (or 3x3) toy example that shows the full use case?  It seems like a useful feature.

**Comment by jklymak** [comment]
When you create an axes, you can set its `sharex` and `sharey` kwargs.  I don't think it'd be unreasonable to make that a public API of an axes object....

For now, I think you can do:

```python
ax._shared_x_axes.join(ax, masterxaxes)
ax._shared_y_axes.join(ax, masteryaxes)
```
I think that is closer to what is wanted, rather than twinning the axes, but maybe I'm misunderstanding..

**Comment by mwaskom** [comment]
> If this can't be achieved w/ the public API, then we should develop a public API to do it, and have tests etc.

I agree. But in the meantime, this change will break one of the most basic seaborn examples, so I'd really also like to advocate for a minor change to the work in #10033 that won't, as far as I understand, undermine any of its actual goals or otherwise allow incorrect specifications.

> @mwaskom can you make a 2x2 (or 3x3) toy example that shows the full use case? It seems like a useful feature.

This is basically what happens in seaborn:

```python
f, axes = plt.subplots(2, 2, figsize=(6, 6), sharex="col", sharey="row")
diag_axes = []
for ax in np.diag(axes):
    if diag_axes:
        diag_ax = ax._make_twin_axes(sharex=ax,
                                     sharey=diag_axes[0],
                                     frameon=False)
    else:
        diag_ax = ax._make_twin_axes(sharex=ax, frameon=False)
    diag_ax.set_axis_off()
    diag_axes.append(diag_ax)
```

**Comment by mwaskom** [comment]
> I think that is closer to what is wanted, rather than twinning the axes, but maybe I'm misunderstanding..

Really the simplest thing would be an `ax.twinx` signature that looks like `def twinx(self, sharey=None)`, but I don't think implementing that will be possible until the change I'm advocating for in the private method is implemented.

**Comment by mwaskom** [comment]
Please let me know if, at least, the usecase is clear, because I really think it's quite straightforward and perhaps being overthought.

**Comment by jklymak** [comment]
BTW, I'll re-open just in way of getting on more people's radars...

**Comment by mwaskom** [comment]
> OTOH, I'm not personally in favour of accomodating this. As @efiring said, seaborn over-specified the sharex argument, for no good reason as far as I can tell. 

The reason is mainly because that's [exactly what `ax.twinx` does](https://github.com/matplotlib/matplotlib/blob/357dad005de2cd4c1cd145482bc41381bd444ddb/lib/matplotlib/axes/_base.py#L4221)!

**Comment by mwaskom** [comment]
> We could put a janky test in _make_twin_axes to see if one of the specified axes is self and allow passing both sharex and sharey, but it really seems this should be patched in seaborn by just dropping the sharex kwargs.

Maybe, maybe not. Given that I'm copying what matplotlib does, I suspect there may be a good reason for it. And in any case, if 2.2 lands with this it will break all existing versions of seaborn with no way to 

> On a larger scale, it seems twinning is the wrong thing to do altogether. If I understand correctly, you don't want a twin, you really want the diagonal axes to have different sharex and sharey than column and row. I think we should/could impliment that. i.e. ax.sharex(listofaxes) and ax.sharey(listofaxes).

seaborn actually does require twin axes because the upper-left hand axes has to have the correct ticks for the other axes in that row. And also it's possible to use the [`PairGrid`](http://seaborn.pydata.org/generated/seaborn.PairGrid.html#seaborn.PairGrid) without having a different axes system for the diagonal plots, and the object doesn't know at the time the original grid of plots is created whether the user is or is not going to draw marginal distributions in the diagonal.

**Comment by dopplershift** [comment]
Speaking as a matplotlib dev with no attachment to the APIs in question, I'm sensitive to the idea that we not break one of the more prominent downstream libraries. So I'm ok with getting in a change to avoid this breakage before we cut 2.2.

I'm also pretty damned irritated that we have to work around this prominent library using an *internal* API; especially since it seems that this downstream library has made no effort whatsoever in getting a public version of this API. (Please correct me if I'm wrong on this.) So I'm also tempted to say that unless a seaborn dev is willing to put some skin in the game and propose a PR to get a public API that serves their purpose, we don't do a damned thing. **THIS** is literally the reason that as a developer of another project, YOU DON'T USE SOMEONE ELSE'S INTERNAL APIS.

**Comment by jklymak** [comment]
OK, fair enough - tracing through all the twin code is difficult.  I did reopen and remilestone so this will at least get discussion before 2.2 drops.  

  1. I'm still not clear why you want twin axes versus just clearing the diagonal axes and resetting the share properties.  But no doubt I'm not following exactly what PairGrid etc are doing.  
  2. Again, the changes implimented in #10033 fix a lot of other critical bugs.  So, if you can figure out how to fix seaborn w/o breaking #10033, I think that PR would be great.

**Comment by mwaskom** [comment]
Thanks @dopplershift for the constructive contributions.

**Comment by mwaskom** [comment]
My proposal is literally to change

```python
if 'sharex' in kwargs and 'sharey' in kwargs:
    raise ValueError("Twinned Axes may share only one axis.")
```

to 

```python
if 'sharex' in kwargs and 'sharey' in kwargs:
    if kwargs["sharex"] is not self and kwargs["sharey"] is not self:
        raise ValueError("Twinned Axes may share only one axis.")
```

**Comment by mwaskom** [comment]
We can also have a conversation about a public API for twinned, shared axes (which I've proposed above), but that shouldn't be a blocker for 2.2. And, as noted, fixing this error check would be a precondition for making that possible.

**Comment by dstansby** [comment]
@mwaskom is it not possible  write (or just copy the old code in `_make_twin_axes`) a method to do what you want to do instead of relying on our private one, and then putting out a patch release of seaborn with that fix?

**Comment by mwaskom** [comment]
No.

**Comment by dstansby** [comment]
Why not?

**Comment by dopplershift** [comment]
That doesn't solve the problem of all the existing installs being broken.

**Comment by jklymak** [comment]
Note that @lkjell took a crack at adding share/unshare in #9923.  I think that is what would be useful for the seaborn use case...

**Comment by jklymak** [comment]
Closed #10622
